tensorflow-core/tensorflow-core-api/src/main/java/org/tensorflow/GraphFunction.java

```diff
@@ -0,0 +1,301 @@
+ *
+ * Functions can also be created from {@link ConcreteFunction}s using {@link #create(ConcreteFunction, boolean)}.
+ */
+public class GraphFunction implements AutoCloseable {
+
+  private final TF_Function nativeHandle;
+  private final PointerScope scope;
+  private final int numInputs;
+  private final int numOutputs;
+
+  GraphFunction(TF_Function nativeHandle, int numInputs, int numOutputs) {
+    this.numInputs = numInputs;
+    this.numOutputs = numOutputs;
+    scope = new PointerScope();
+    this.nativeHandle = nativeHandle;
+    scope.attach(nativeHandle);
+  }
+
+  /**
+   * This causes native errors when I use it (Linux GPU, 6.1 CC), but I'm leaving it because how to enable XLA JIT is
+   * extremely non-obvious.
+   *
+   * Causes {@code OP_REQUIRES failed at xla_ops.cc:363 : Not found: could not find registered platform with id:
+   * 0x7f75af03e6e8} (it's a warning, but the resulting TF_Status fails).
+   */
+  private void makeJit() {
+    try (PointerScope scope = new PointerScope()) {
+      byte[] bytes = AttrValue.newBuilder().setB(true).build().toByteArray();
+      BytePointer trueValue = new BytePointer(bytes);
+
+      TF_Status status1 = TF_Status.newStatus();
+      TF_FunctionSetAttrValueProto(nativeHandle(), "_XlaMustCompile", trueValue, bytes.length, status1);
+      status1.throwExceptionIfNotOK();
+
+      TF_Status status2 = TF_Status.newStatus();
+      TF_FunctionSetAttrValueProto(nativeHandle(), "_noinline", trueValue, bytes.length, status2);
+      status2.throwExceptionIfNotOK();
+    }
+  }
+
+  private static TF_Operation outputHandle(Operand<?> operand) {
+    if (operand == null) {
+      throw new NullPointerException("Can't get output handle for null operand");
+    }
+
+    Pointer handle = operand.asOutput().getUnsafeNativeHandle();
+    if (handle.isNull()) {
+      throw new NullPointerException("Native handle of operand is null, has it been closed?");
+    }
+
+    if (!(handle instanceof TF_Operation)) {
+      throw new IllegalArgumentException("Operand was not a graph operand");
+    }
+
+    return (TF_Operation) handle;
+  }
+
+  private static TF_Output resolveToOutput(List<Operand<?>> operands) {
+    TF_Output handles = new TF_Output(operands.size());
+    for (int i = 0; i < operands.size(); i++) {
+      Operand<?> input = operands.get(i);
+      //TODO body.checkInput(input)
+      TF_Operation handle = outputHandle(input);
+      handles.position(i).oper(handle).index(input.asOutput().index());
+    }
+    handles.position(0);
+    return handles;
+  }
+
+  /**
+   * Create a {@link GraphFunction} with positional argument and output resolution. Includes all ops in {@code body},
+   * {@code inputs} and {@code outputs} must be the inputs and outputs of {@code body}.
+   *
+   * @param body the graph to use as the function body
+   * @param name the name of the function
+   * @param appendHash whether to append the has of the signature to the name
```

Comment:
".. the has*h*..."

btw, just curious, why do we let the user decide to append a hash or not to the signature name? Can we take this decision for him?

Reply:
I'd lean in favor of always appending it, but I could see a case where you want to create a function with a specific name, so I left it in.  It's definitely necessary most of the time to avoid name collisions for functions with different arguments (aka overloading support).